commit 2b33d877cb03
Author: Andreas Pehrson <pehrsons@mozilla.com>
Date:   Mon Mar 19 15:52:36 2018 +0100

    Bug 1444541 - Don't stop recorder on pagehide during frameswap. r=smaug, a=jcristau
    
    MozReview-Commit-ID: AZec0yXM1FI
    
    --HG--
    extra : source : bcdb264028e930756222a756e83bdca8574278e2
---
 dom/media/MediaRecorder.cpp | 18 +++++++++++++++---
 dom/media/moz.build         |  1 +
 2 files changed, 16 insertions(+), 3 deletions(-)

diff --git dom/media/MediaRecorder.cpp dom/media/MediaRecorder.cpp
index 60fe1ea0dae2..a229e4c38d73 100644
--- dom/media/MediaRecorder.cpp
+++ dom/media/MediaRecorder.cpp
@@ -30,6 +30,7 @@
 #include "nsCharSeparatedTokenizer.h"
 #include "nsContentTypeParser.h"
 #include "nsContentUtils.h"
+#include "nsDocShell.h"
 #include "nsError.h"
 #include "nsIDocument.h"
 #include "nsIPermissionManager.h"
@@ -1699,9 +1700,20 @@ MediaRecorder::NotifyOwnerDocumentActivityChanged()
   nsIDocument* doc = window->GetExtantDoc();
   NS_ENSURE_TRUE_VOID(doc);
 
-  LOG(LogLevel::Debug, ("MediaRecorder %p document IsActive %d isVisible %d\n",
-                     this, doc->IsActive(), doc->IsVisible()));
-  if (!doc->IsActive() || !doc->IsVisible()) {
+  bool inFrameSwap = false;
+  if (nsDocShell* docShell = static_cast<nsDocShell*>(doc->GetDocShell())) {
+    inFrameSwap = docShell->InFrameSwap();
+  }
+
+  LOG(LogLevel::Debug, ("MediaRecorder %p NotifyOwnerDocumentActivityChanged "
+                        "IsActive=%d, "
+                        "IsVisible=%d, "
+                        "InFrameSwap=%d",
+                        this,
+                        doc->IsActive(),
+                        doc->IsVisible(),
+                        inFrameSwap));
+  if (!doc->IsActive() || !(inFrameSwap || doc->IsVisible())) {
     // Stop the session.
     ErrorResult result;
     Stop(result);
diff --git dom/media/moz.build dom/media/moz.build
index 194c73fd2f4c..4b1ca538f927 100644
--- dom/media/moz.build
+++ dom/media/moz.build
@@ -312,6 +312,7 @@ EXTRA_JS_MODULES.media += [
 
 LOCAL_INCLUDES += [
     '/caps',
+    '/docshell/base',
     '/dom/base',
     '/layout/generic',
     '/layout/xul',
